node/core/approval-voting/src/backend.rs

```diff
@@ -0,0 +1,305 @@
+}
+
+pub(super) struct DbBackend {
+	inner: Arc<dyn KeyValueDB>,
+	pub(super) config: Config,
+}
+
+impl DbBackend {
+	/// Create a new [`DbBackend`] with the supplied key-value store and
+	/// config.
+	pub fn new(db: Arc<dyn KeyValueDB>, config: Config) -> Self {
+		DbBackend {
+			inner: db,
+			config,
+		}
+	}
+}
+
+impl Backend for DbBackend {
+	fn load_block_entry(
+		&self,
+		block_hash: &Hash,
+	) -> SubsystemResult<Option<BlockEntry>> {
+		ops::load_block_entry(&*self.inner, &self.config, block_hash)
+			.map(|e| e.map(Into::into))
+	}
+
+	fn load_candidate_entry(
+		&self,
+		candidate_hash: &CandidateHash,
+	) -> SubsystemResult<Option<CandidateEntry>> {
+		ops::load_candidate_entry(&*self.inner, &self.config, candidate_hash)
+			.map(|e| e.map(Into::into))
+	}
+
+	fn load_blocks_at_height(
+		&self,
+		block_height: &BlockNumber
+	) -> SubsystemResult<Vec<Hash>> {
+		ops::load_blocks_at_height(&*self.inner, &self.config, block_height)
+	}
+
+	fn load_all_blocks(&self) -> SubsystemResult<Vec<Hash>> {
+		ops::load_all_blocks(&*self.inner, &self.config)
+	}
+
+	fn load_stored_blocks(&self) -> SubsystemResult<Option<StoredBlockRange>> {
+		ops::load_stored_blocks(&*self.inner, &self.config)
+	}
+
+	/// Atomically write the list of operations, with later operations taking precedence over prior.
+	fn write<I>(&mut self, ops: I) -> SubsystemResult<()>
+		where I: IntoIterator<Item = BackendWriteOp>
+	{
+		let mut tx = DBTransaction::new();
+		for op in ops {
+			match op {
+				BackendWriteOp::WriteStoredBlockRange(stored_block_range) => {
+					tx.put_vec(
+						self.config.col_data,
+						&STORED_BLOCKS_KEY,
+						stored_block_range.encode(),
+					);
+				}
+				BackendWriteOp::WriteBlocksAtHeight(h, blocks) => {
+					tx.put_vec(
+						self.config.col_data,
+						&blocks_at_height_key(h),
+						blocks.encode(),
+					);
+				}
+				BackendWriteOp::DeleteBlocksAtHeight(h) => {
+					tx.delete(
+						self.config.col_data,
+						&blocks_at_height_key(h),
+					);
+				}
+				BackendWriteOp::WriteBlockEntry(block_entry) => {
+					let block_entry: approval_db::v1::BlockEntry = block_entry.into();
+					tx.put_vec(
+						self.config.col_data,
+						&block_entry_key(&block_entry.block_hash),
+						block_entry.encode(),
+					);
+				}
+				BackendWriteOp::DeleteBlockEntry(hash) => {
+					tx.delete(
+						self.config.col_data,
+						&block_entry_key(&hash),
+					);
+				}
+				BackendWriteOp::WriteCandidateEntry(candidate_entry) => {
+					let candidate_entry: approval_db::v1::CandidateEntry = candidate_entry.into();
+					tx.put_vec(
+						self.config.col_data,
+						&candidate_entry_key(&candidate_entry.candidate.hash()),
+						candidate_entry.encode(),
+					);
+				}
+				BackendWriteOp::DeleteCandidateEntry(candidate_hash) => {
+					tx.delete(
+						self.config.col_data,
+						&candidate_entry_key(&candidate_hash),
+					);
+				}
+			}
+		}
+
+		self.inner.write(tx)?;
+		Ok(())
```

Comment:
`Ok(())` here is redundant if the `?;` is omitted from the above line, right?

Reply:
The difference is the current impl does a implicit type conversion, if the returned Error types are the same, it indeed can be omitted.